kernel/src/main/java/org/kframework/compile/ResolveFun.java

```diff
@@ -101,15 +106,15 @@ public K apply(KApply k) {
             body = k.items().get(0);
             arg = k.items().get(1);
           }
-          if (arg instanceof KVariable) {
-            nameHint1 = ((KVariable) arg).name();
-          } else if (arg instanceof KApply
-              && ((KApply) arg).klabel().name().startsWith("#SemanticCastTo")
-              && ((KApply) arg).items().get(0) instanceof KVariable) {
-            nameHint1 = ((KVariable) ((KApply) arg).items().get(0)).name();
+          if (arg instanceof KVariable var) {
+            nameHint1 = var.name();
+          } else if (arg instanceof KApply app
+              && app.klabel().name().startsWith("#SemanticCastTo")
+              && app.items().get(0) instanceof KVariable) {
+            nameHint1 = ((KVariable) app.items().get(0)).name();
```

Comment:
I don't think that's enough. You may have something like `(Var):Sort1` or `Var:Sort1:Sort2`
Weird but it can happen. The proper way to check if you have a variable under a specific term is with a visitor.
Maybe you can use something like this: https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/compile/ResolveSemanticCasts.java#L104
and remove the code duplication from `isLHSTotal`.

Reply:
Great, that's what I was looking for - will implement in terms of a visitor. Thanks Radu!

Reply:
I have implemented the corrected behaviour in terms of a direct recursion rather than a visitor; the core logic is the same but I think the code is less noisy as a result. Happy to refactor if you'd prefer @radumereuta

Reply:
(either way, the duplication is resolved)